File: benchmark/rabbitmq_producer.py
import time
import pika
import json
from config import MESSAGE_COUNT, TOPIC_NAME, PAYLOAD

def run():
    connection = pika.BlockingConnection(
        pika.ConnectionParameters(host='localhost', port=5672, credentials=pika.PlainCredentials('user', 'password'))
    )
    channel = connection.channel()

    channel.queue_declare(queue=TOPIC_NAME)

    print(f"RabbitMQ Producer: Starting to produce {MESSAGE_COUNT} messages...")
    start = time.time()
    
    payload_bytes = json.dumps(PAYLOAD).encode('utf-8')

    for i in range(MESSAGE_COUNT):
        channel.basic_publish(exchange='', routing_key=TOPIC_NAME, body=payload_bytes)
        if i % 1000 == 0:
            print(f"Produced {i} messages", end='\r')

    connection.close()
    end = time.time()
    print(f"\nRabbitMQ Producer: {MESSAGE_COUNT} messages in {end - start:.4f} seconds")

if __name__ == "__main__":
    run()

File: typescript/benchmark/rabbitmq_producer.ts
import * as amqp from 'amqplib';
import { MESSAGE_COUNT, TOPIC_NAME, PAYLOAD } from './config';
import { saveResult } from './report';

async function run() {
    const conn = await amqp.connect('amqp://user:password@localhost:5672');
    const ch = await conn.createChannel();
    await ch.assertQueue(TOPIC_NAME);

    console.log(`RabbitMQ Producer (TS): Starting to produce ${MESSAGE_COUNT} messages...`);
    const start = Date.now();
    const payload = Buffer.from(JSON.stringify(PAYLOAD));

    for (let i = 1; i <= MESSAGE_COUNT; i++) {
        const ok = ch.sendToQueue(TOPIC_NAME, payload);
        if (!ok) {
            await new Promise<void>(resolve => ch.once('drain', resolve));
        }
        
        if (i % 100000 === 0) {
            const duration = (Date.now() - start) / 1000;
            console.log(`RabbitMQ Producer (TS): Reached ${i} messages in ${duration.toFixed(4)}s`);
            saveResult("RabbitMQ (TS)", "Producer", i, duration);
        }
    }

    await ch.close();
    await conn.close();
    
    const duration = (Date.now() - start) / 1000;
    console.log(`\nRabbitMQ Producer (TS): ${MESSAGE_COUNT} messages in ${duration.toFixed(4)} seconds`);
    saveResult("RabbitMQ (TS)", "Producer", MESSAGE_COUNT, duration);
}

run().catch(console.error);
